test/unit/puppetlabs/services/ca/certificate_authority_core_test.clj

```diff
@@ -119,6 +119,44 @@
           response (ring-app request)]
       (is (= version-number (get-in response [:headers "X-Puppet-Version"]))))))
 
+(deftest handle-delete-certificate-request!-test
+  (let [settings      (assoc (testutils/ca-sandbox! cadir)
+                             :allow-duplicate-certs true
+                             :autosign false)
+        subject       "foo-agent"
+        csr-stream    (io/input-stream (test-pem-file "foo-agent-csr.pem"))
+        expected-path (ca/path-to-cert-request (:csrdir settings) subject)
+        assert-proper
+          (fn
+            [response code matcher log-level]
+            (do
+             (is (= code (:status response)))
+             (is (re-matches matcher (:body response)))
+             (is (= "text/plain" (get-in response [:headers "Content-Type"])))
+             (is (logged? matcher log-level))))]
+    (logutils/with-test-logging
+      (testing "Successful CSR deletion"
+        (try
+          (handle-put-certificate-request! subject csr-stream settings)
+          (is (true? (fs/exists? expected-path)))
+          (let [response (handle-delete-certificate-request! subject settings)
+                msg-matcher (re-pattern (str "Deleted .* for " subject ".*"))]
+            (is (false? (fs/exists? expected-path)))
+            (assert-proper response 204 msg-matcher :debug))
+          (finally
+            (fs/delete expected-path))))
+      (testing "Attempted deletion of a non-existant CSR"
+        (let [response (handle-delete-certificate-request! subject settings)
+              msg-matcher (re-pattern (str "No cert.* " subject " at .*"))]
+          (is (false? (fs/exists? expected-path)))
+          (assert-proper response 404 msg-matcher :info)))
+      (testing "Error during deletion of a CSR"
+        (let [response
+              (with-redefs [me.raynes.fs/exists? (constantly true)]
```

Comment:
So I'm able to reproduce this behavior in the repl bypassing me.raynes.fs:
```
16:00:24 hobo:pems justin (bug/master/SERVER-977_add_cert_request_delete *%) ‖ ls -l foo-agent-csr.pem
-rw-r--r--  1 justin  staff  985 Apr  5 15:24 foo-agent-csr.pem
```

```clj
user=> (java.nio.file.Files/isWritable (java.nio.file.Paths/get (.toURI fp)))
true
user=> (.canWrite fp)
true
user=> (.setWritable fp false)
true
user=> (.canWrite fp)
false
user=> (java.nio.file.Files/isWritable (java.nio.file.Paths/get (.toURI fp)))
false
user=> (java.nio.file.Files/delete (java.nio.file.Paths/get (.toURI fp)))
nil
user=> (.canWrite fp)
false
```
```
16:00:28 hobo:pems justin (bug/master/SERVER-977_add_cert_request_delete *%) ‖ ls -l foo-agent-csr.pem
ls: foo-agent-csr.pem: No such file or directory
```


I don't know if my understanding of the way Java or how filesystems work is broken. Or just my filesystem and Java are having a hard time playing nicely.